src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs

```diff
@@ -0,0 +1,138 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using System.Runtime.InteropServices;
+
+namespace System.Reflection.Emit
+{
+    internal sealed class ModuleBuilderImpl : ModuleBuilder
+    {
+        private readonly Assembly _coreAssembly;
+        private readonly string _name;
+        private readonly Dictionary<string, Type> _coreTypes = new();
+        private readonly Dictionary<Assembly, AssemblyReferenceHandle> _assemblyRefStore = new();
+        private readonly Dictionary<Type, TypeReferenceHandle> _typeRefStore = new();
+        private readonly List<TypeBuilderImpl> _typeDefStore = new();
+        private int _nextMethodDefRowId = 1;
+        private int _nextFieldDefRowId = 1;
+
+        internal ModuleBuilderImpl(string name, Assembly coreAssembly)
+        {
+            _coreAssembly = coreAssembly;
+            _name = name;
+        }
+
+        internal Type GetTypeFromCoreAssembly(string name)
+        {
+            Type? type;
+
+            // TODO: Use Enum as the key for perf
+            if (!_coreTypes.TryGetValue(name, out type))
+            {
+#pragma warning disable IL2026 // Members annotated with 'RequiresUnreferencedCodeAttribute' require dynamic access otherwise can break functionality when trimming application code
+                type = _coreAssembly.GetType(name, throwOnError: true)!;
+#pragma warning restore IL2026
+                _coreTypes.Add(name, type);
+            }
+
+            return type;
+        }
+
+        internal void AppendMetadata(MetadataBuilder metadata)
+        {
+            // Add module metadata
+            metadata.AddModule(
+                generation: 0,
+                moduleName: metadata.GetOrAddString(ScopeName),
```

Comment:
```suggestion
                moduleName: metadata.GetOrAddString(_name),
```
Nit: We can use the name field directly